src/qt/guiutil.cpp

```diff
@@ -376,23 +377,93 @@ bool WindowContextHelpButtonHintFilter::eventFilter (QObject *obj, QEvent *event
     return QObject::eventFilter (obj, event);
 }
 
+
+struct AutoStartupArguments
+{
+    std::string link_name_suffix;
+    boost::filesystem::path data_dir;
+    std::string arguments;
+};
+
+AutoStartupArguments GetAutoStartupArguments()
+{
+    // This helper function checks for the presence of certain startup arguments
+    // to the current running instance that should be relevant for automatic restart
+    // (currently testnet, datadir, scraper, explorer, usenewnn). It adds -testnet
+    // to the link name as a suffix if -testnet is specified otherwise adds mainnet,
+    // and then adds the other three as arguments if they were specified for the
+    // running instance. This allows two different automatic startups, one for
+    // mainnet, and the other for testnet, and each of them can have different datadir,
+    // scraper, explorer, and/or usenewnn arguments.
+
+    AutoStartupArguments result;
+
+    // We can't use GetDataDir() here because it appends /testnet for
+    // testnet, which is correct for internal use, but not what we
+    // need for the command line argument. I think the fNetSpecific flag
+    // for GetDataDir is not coded right.
+    // TODO: Review GetDataDir() fix and replace this.
+    if (mapArgs.count("-datadir"))
+    {
+            result.data_dir = fs::system_complete(mapArgs["-datadir"]);
+            if (!fs::is_directory(result.data_dir))
+            {
+                result.data_dir = "";
+            }
+    }
+
+    result.arguments = "-min";
+
+    if (fTestNet)
+    {
+        result.link_name_suffix += "-testnet";
+        result.arguments += " -testnet";
+    }
+    else
+    {
+        result.link_name_suffix += "-mainnet";
+    }
+
+    if (GetBoolArg("-scraper")) result.arguments += " -scraper";
+
+    if (GetBoolArg("-explorer")) result.arguments += " -explorer";
+
+    if (GetBoolArg("-usenewnn")) result.arguments += " -usenewnn";
```

Comment:
```suggestion
for (const auto& flag : { "-scraper", "-explorer", "-usenewnn" }) 
{
    if (GetBoolArg(flag)) 
    {
        (result.arguments += " ") += flag;
        result.link_name_suffix += flag;
    }
}
```
What do you think about appending these to the shortcut name to avoid clashes for cases when two wallets run on the same host? It may make it more clear what the shortcut is going to open as well. Not that this will happen in practice very often.

Reply:
I was thinking about these, but not with these. If we are going to to that, to support an arbitrary wallet instance, I was thinking about doing a truncated hash of the wallet exe, testnet, and datadir arguments. Something like the first eight of an SHA256 should be adequate.

Reply:
One other comment... the loop is shorter but on the hairy edge of warranted for three of them, and less obvious. I suppose it is easier to extend if we put additional arguments we want to include.

Reply:
Hmm. I like the hash idea, but maybe we're making this too complicated for a feature that fewer than 10 people will really use. Perhaps we avoid ambiguity if we don't include the flags at all, and power users can set up their environments as needed. 

I agree&mdash;the loop is a rather heavy abstraction for this simple code. I just figured that using the flag constants now in three places in the suggested change tipped the threshold. I do not prefer one style over the other.

Reply:
I am fine with the change. :)